File: tests/unit/deflate.test.ts
import { describe, expect, it } from 'vitest'
import { compress, decompress, type CompressedResult } from '../../src/codec/compression/deflate'

// Deterministic PRNG so the "random" fixture is reproducible across runs.
function mulberry32(seed: number): () => number {
  let a = seed >>> 0
  return () => {
    a += 0x6d2b79f5
    let t = a
    t = Math.imul(t ^ (t >>> 15), t | 1)
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61)
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }
}

function randomBytes(length: number, seed: number): Uint8Array {
  const rand = mulberry32(seed)
  const bytes = new Uint8Array(length)
  for (let i = 0; i < length; i++) {
    bytes[i] = Math.floor(rand() * 256)
  }
  return bytes
}

function repeatedByte(byte: number, length: number): Uint8Array {
  const bytes = new Uint8Array(length)
  bytes.fill(byte)
  return bytes
}

function mixedText(length: number, seed: number): Uint8Array {
  const rand = mulberry32(seed)
  const words = ['the', 'quick', 'brown', 'fox', 'jumps', 'over', 'lazy', 'dog', 'qr', 'transfer']
  const bytes = new Uint8Array(length)
  for (let i = 0; i < length; i++) {
    if (i % 32 === 0) {
      // `rand() < 1` bounds the index, but noUncheckedIndexedAccess can't see it.
      const word = words[Math.floor(rand() * words.length)] ?? 'qr'
      for (let j = 0; j < word.length && i + j < length; j++) {
        bytes[i + j] = word.charCodeAt(j)
      }
      i += word.length - 1
    } else {
      bytes[i] = 32 + Math.floor(rand() * 95) // printable ASCII
    }
  }
  return bytes
}

function expectRoundTrip(input: Uint8Array): void {
  const result = compress(input)
  const restored = decompress(result)
  expect(restored.length).toBe(input.length)
  expect(restored).toEqual(input)
}

function expectSameBytes(a: Uint8Array, b: Uint8Array): void {
  expect(a.length).toBe(b.length)
  expect(a).toEqual(b)
}

describe('compress / decompress', () => {
  it('round-trips an empty input and reports compressed=false', () => {
    const input = new Uint8Array(0)
    const result = compress(input)
    expect(result.compressed).toBe(false)
    expect(result.data.length).toBe(0)
    expectRoundTrip(input)
  })

  it('round-trips a single byte', () => {
    const input = new Uint8Array([42])
    expectRoundTrip(input)
  })

  it('compresses highly compressible repeated data and round-trips', () => {
    const input = repeatedByte('a'.charCodeAt(0), 10 * 1024)
    const result = compress(input)
    expect(result.compressed).toBe(true)
    expect(result.data.length).toBeLessThan(input.length)
    expectRoundTrip(input)
  })

  it('skips compression when deflate would not shrink the data', () => {
    const input = randomBytes(10 * 1024, 0xc0ffee)
    const result = compress(input)
    expect(result.compressed).toBe(false)
    expect(result.data).toBe(input)
    expectRoundTrip(input)
  })

  it('round-trips 100KB of mixed text', () => {
    const input = mixedText(100 * 1024, 7)
    expectRoundTrip(input)
  })

  it('is deterministic: identical input produces identical compressed bytes', () => {
    const input = repeatedByte('a'.charCodeAt(0), 10 * 1024)
    const first = compress(input)
    const second = compress(input)
    expect(first.compressed).toBe(second.compressed)
    if (first.compressed && second.compressed) {
      expectSameBytes(first.data, second.data)
    }
  })

  it('decompress returns the original reference on the non-compressed path', () => {
    const input = randomBytes(10 * 1024, 0xdeadbeef)
    const result: CompressedResult = { data: input, compressed: false }
    const restored = decompress(result)
    expect(restored).toBe(input)
    expectSameBytes(restored, input)
  })

  it('decompress inflates the compressed path back to the original bytes', () => {
    const input = repeatedByte('a'.charCodeAt(0), 10 * 1024)
    const result = compress(input)
    expect(result.compressed).toBe(true)
    const restored = decompress(result)
    expectSameBytes(restored, input)
  })
})

File: flutter_app/core/test/codec/deflate_test.dart
import 'dart:io';
import 'dart:math';
import 'dart:typed_data';

import 'package:qr_transfer_core/codec/deflate.dart';
import 'package:test/test.dart';

void main() {
  group('compress/decompress round-trip', () {
    test('empty input is reported uncompressed and round-trips empty', () {
      final input = Uint8List(0);

      final result = compress(input);

      expect(result.compressed, isFalse);
      expect(decompress(result), isEmpty);
    });

    test('single byte round-trips byte-identical', () {
      final input = Uint8List.fromList([0x42]);

      final result = compress(input);

      expect(result.compressed, isFalse);
      expect(decompress(result), equals(input));
    });

    test('10KB repeated a compresses (true, smaller) and round-trips', () {
      final input = Uint8List(10 * 1024)..fillRange(0, 10 * 1024, 0x61);

      final result = compress(input);

      expect(result.compressed, isTrue);
      expect(result.data.length, lessThan(input.length));
      expect(decompress(result), equals(input));
    });

    test('10KB seeded random is kept uncompressed and round-trips', () {
      final rng = Random(42);
      final input = Uint8List.fromList(
        List.generate(10 * 1024, (_) => rng.nextInt(256)),
      );

      final result = compress(input);

      expect(result.compressed, isFalse);
      expect(result.data, equals(input));
      expect(decompress(result), equals(input));
    });

    test('100KB mixed text round-trips byte-identical', () {
      final rng = Random(7);
      final input = Uint8List.fromList(
        List.generate(100 * 1024, (_) => 32 + rng.nextInt(95)),
      );

      final result = compress(input);
      expect(decompress(result), equals(input));
    });
  });

  group('determinism', () {
    test('same input compresses to identical bytes on every call', () {
      final input = Uint8List(10 * 1024)..fillRange(0, 10 * 1024, 0x61);

      final first = compress(input).data;
      final second = compress(input).data;

      expect(second, equals(first));
    });
  });

  group('pako byte-parity (cross-engine proof)', () {
    test(
      'Dart zlib level-6 output equals pako deflate fixture byte-for-byte',
      () {
        final original = File(
          'test/fixtures/text-256k/original.bin',
        ).readAsBytesSync();
        final pakoOutput = File(
          'test/fixtures/text-256k/deflate.bin',
        ).readAsBytesSync();

        final result = compress(original);

        expect(result.compressed, isTrue);
        expect(result.data, equals(pakoOutput));
      },
    );

    test('pako deflate fixture inflates back to the original bytes', () {
      final original = File(
        'test/fixtures/text-256k/original.bin',
      ).readAsBytesSync();
      final pakoOutput = File(
        'test/fixtures/text-256k/deflate.bin',
      ).readAsBytesSync();

      final restored = decompress(
        CompressedResult(data: pakoOutput, compressed: true),
      );

      expect(restored, equals(original));
    });
  });

  group('fixtures', () {
    test('random-64k deflate.bin is the uncompressed original '
        '(PWA skip-if-not-smaller)', () {
      final original = File(
        'test/fixtures/random-64k/original.bin',
      ).readAsBytesSync();
      final deflateBin = File(
        'test/fixtures/random-64k/deflate.bin',
      ).readAsBytesSync();

      expect(deflateBin, equals(original));

      final result = compress(original);
      expect(result.compressed, isFalse);
      expect(result.data, equals(original));
    });
  });
}
